### PR TITLE
Recipe Rewrite (API Only)

### PR DESCRIPTION
**Pull request by SchnTgaiSpock**
## Description
This PR supersedes #4078

### Description from #4078
Hey guys, this is my attempt at the recipe rewrite.

I have written up an [ADR](https://github.com/SchnTgaiSpock/Slimefun4/blob/api/recipe-rewrite-3/docs/adr/0002-recipe-rewrite.md) that goes over how the new API currently works, but in short:
- Recipe system sucks
- Can't translate recipes in the dispenser (like crafting Copper Wire on the top row)
- Items can only have one* recipe (mostly)
- Can't have true shapeless recipes

Any feedback on the API or discussion around the new recipe system in general would be greatly appreciated!

## Proposed changes
- Adds a new recipe API which:
  - Allows recipe inputs with multiple options (e.g. Vanilla/Slimefun Copper)
  - Allows true shaped/shapeless recipes (does not replace the old versions)
  - Allows machines and workstations to search recipes more efficiently
  
## Related Issues (if applicable)
N/A

## Checklist
- [ ] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [ ] I have made sure that the proposed changes do not break compatibility across the supported Minecraft versions (1.16.* - 1.20.*).
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [x] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

**Review by J3fftw1 — CHANGES_REQUESTED** [review]
Can we get unit tests on this?

**Comment by SchnTgaiSpock** [comment]
yup thats my prio rn

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/Slimefun.java [hunk 1]
```diff
@@ -795,7 +803,7 @@ private static void validateInstance() {
 
     /**
      * This method returns out {@link MinecraftRecipeService} for Slimefun.
-     * This service is responsible for finding/identifying {@link Recipe Recipes}
+     * This service is responsible for finding/identifying {@link org.bukkit.inventory.Recipe Recipes}
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/Slimefun.java:806]: Please import this

**Comment by SchnTgaiSpock** [comment]
unit tests should be done, if i missed something lmk

**Comment by JustAHuman-xD** [comment]
Might be worth waiting for the item stack component changes before doing much more with this

**Pull request closed without merging**